modules/network/src/main/java/org/apache/ignite/internal/network/DefaultMessagingService.java

```diff
@@ -705,4 +724,18 @@ protected void onStoppingInitiated() {
     private static InetSocketAddress createResolved(NetworkAddress address) {
         return new InetSocketAddress(address.host(), address.port());
     }
+
+    private InetSocketAddress getFromCacheOrCreateResolved(ClusterNode recipientNode) {
+        assert recipientNode.name() != null : "Node has not been added to the topology: " + recipientNode.id();
+
+        InetSocketAddress address = recipientInetAddrByNodeId.compute(recipientNode.id(), (nodeId, inetSocketAddress) -> {
```

Comment:
We can theoretically resurrect the cache if we send a message to offline node, how are we avoiding it?
Why don't we cache address instance inside of `ClusterNode` like we did in Ignite 2, why map instead?

Reply:
> We can theoretically resurrect the cache if we send a message to offline node, how are we avoiding it?

If I understand correctly, when a node becomes offline, it ends up in the “list” of stale nodes, by performing the check below we will avoid being filled with such nodes.
```
if (staleIdDetector.isIdStale(nodeId)) {
    return null;
}
```

> Why don't we cache address instance inside of ClusterNode like we did in Ignite 2, why map instead?

`org.apache.ignite.network.ClusterNode` is a public API, I don’t think the user needs to give the `InetSocketAddress`, so an internal map is used. What do you think?

Reply:
> by performing the check below we will avoid being filled with such nodes.

This is cool, I completely forgot about it!
> is a public API

Ok, I see, although it's an interface and we could hide resolved address in the implementation. Anyway, we shouldn't use the same classes for private and public API, makes no sense to me, but it's not related to your changes.

> What do you think?
I agree with you, map will do it for now, thank you!